_posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd

```diff
@@ -0,0 +1,492 @@
+```
+
+Inspection could be performed using any group element. I'll pick rotation by $\pi/2$:
+
+```{r}
+all <- iterate(r2_act$testing_elements)
+g1 <- all[[2]]
+g1
+```
+
+Just for fun, let's see how we can -- literally -- come whole circle by letting this element act on the input tensor four times:
+
+```{r}
+all <- iterate(r2_act$testing_elements)
+g1 <- all[[2]]
+
+x1 <- x$transform(g1)
+x1$tensor
+x2 <- x1$transform(g1)
+x2$tensor
+x3 <- x2$transform(g1)
+x3$tensor
+x4 <- x3$transform(g1)
+x4$tensor
+```
+
+```         
+tensor([[[[ 1.,  1.,  1.,  1.],
+          [ 0.,  1.,  2.,  3.],
+          [ 0.,  1.,  4.,  9.],
+          [ 0.,  1.,  8., 27.]]]])
+          
+tensor([[[[ 1.,  3.,  9., 27.],
+          [ 1.,  2.,  4.,  8.],
+          [ 1.,  1.,  1.,  1.],
+          [ 1.,  0.,  0.,  0.]]]])
+          
+tensor([[[[27.,  8.,  1.,  0.],
+          [ 9.,  4.,  1.,  0.],
+          [ 3.,  2.,  1.,  0.],
+          [ 1.,  1.,  1.,  1.]]]])
+          
+tensor([[[[ 0.,  0.,  0.,  1.],
+          [ 1.,  1.,  1.,  1.],
+          [ 8.,  4.,  2.,  1.],
+          [27.,  9.,  3.,  1.]]]])
+```
+
+You see that at the end, we are back at the original "image".
+
+Now, for equivariance. We could first apply a rotation, then convolve. Rotate:
+
+```{r}
+x_rot <- x$transform(g1)
+x_rot$tensor
+```
+
+This is the first in the above list of four tensors. -- Convolve:
+
+```{r}
+y <- conv(x_rot)
+y$tensor
+```
+
+```         
+tensor([[[[ 1.1955,  1.7110],
+          [-0.5166,  1.0665]],
+
+         [[-0.0905,  2.6568],
+          [-0.3743,  2.8144]],
+
+         [[ 5.0640, 11.7395],
+          [ 8.6488, 31.7169]],
+
+         [[ 2.3499,  1.7937],
+          [ 4.5065,  5.9689]]]], grad_fn=<ConvolutionBackward0>)
+```
+
+Alternatively, we can do the convolution first, then rotate its output. Convolve:
+
+```{r}
+y_conv <- conv(x)
+y_conv$tensor
+```
+
+```         
+tensor([[[[-0.3743, -0.0905],
+          [ 2.8144,  2.6568]],
+
+         [[ 8.6488,  5.0640],
+          [31.7169, 11.7395]],
+
+         [[ 4.5065,  2.3499],
+          [ 5.9689,  1.7937]],
+
+         [[-0.5166,  1.1955],
+          [ 1.0665,  1.7110]]]], grad_fn=<ConvolutionBackward0>)
+```
+
+Rotate:
+
+```{r}
+y <- y_conv$transform(g1)
+y$tensor
+```
+
+```         
+tensor([[[[ 1.1955,  1.7110],
+          [-0.5166,  1.0665]],
+
+         [[-0.0905,  2.6568],
+          [-0.3743,  2.8144]],
+
+         [[ 5.0640, 11.7395],
+          [ 8.6488, 31.7169]],
+
+         [[ 2.3499,  1.7937],
+          [ 4.5065,  5.9689]]]])
+```
+
+Indeed, final results are the same.
+
+At this point, we know how to employ group-equivariant convolutions. The final step is to compose the network.
+
+## A group-equivariant neural network
+
+Basically, we have two questions to answer. The first concerns the non-linearities. This is a potentially intricate topic, but as long as we stay with point-wise operations (such as that performed by ReLU) equivariance is given intrinsically.
```

Comment:
> we have two questions to answer.

I think you delve deeply into the first before you get to the second, so you might want to either say what they both are now, or start this with something to the effect of “The first question we have to answer concerns the non-linearities."

Reply:
Having read the whole section, I'm not totally sure I get what that second question is 😬

Reply:
Oh yes that's correct! I'll do as you say, and name the second immediately.